ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java

```diff
@@ -938,6 +949,339 @@ private void updateClusterMapInHelix(boolean startValidatingClusterManager) thro
+   * partitions and find corresponding resources for them if possible.
+   * Before calling this method, we should already verify that each instance has only one assigned resource. So if a
+   * partition's instances are already assigned to the same resource, then this partition would be assigned to this resource
+   * as well. If a partition's instances are assigned to different resources, then this method would fail.
+   * @param dcName The name of the datacenter
+   * @param partitionsToInstancesInDc The map of partition layout.
+   * @param instanceNameToOneResource The map from instance name to its corresponding assigned resource id
+   * @param resourceIdToInstances The map from resource id to a list of instance names that are assigned to this resource
+   * @param resourceIdToIdealState The map from resource id to IdealState
+   */
+  private void updateIdealStateWithNewPartitionsForFullAuto(String dcName,
+      Map<String, Set<String>> partitionsToInstancesInDc, Map<String, Integer> instanceNameToOneResource,
+      Map<Integer, Set<String>> resourceIdToInstances, Map<Integer, IdealState> resourceIdToIdealState) {
+    HelixAdmin dcAdmin = adminForDc.get(dcName);
+
+    // Find the partitions whose instances already exist in resources and assign those partitions to the corresponding resources.
+    int round = 1;
+    boolean allPartitionsDontHaveResources;
+    while (round >= 0) {
+      allPartitionsDontHaveResources = true;
+      for (Map.Entry<String, Set<String>> ent : partitionsToInstancesInDc.entrySet()) {
+        Set<String> instanceNames = ent.getValue();
+        Set<Integer> resources = instanceNames.stream()
+            .map(in -> instanceNameToOneResource.get(in))
+            .filter(Objects::nonNull)
+            .collect(Collectors.toSet());
+        if (resources.isEmpty()) {
+          // Partition doesn't belong to any host that are already in the helix
+        } else if (resources.size() == 1) {
+          allPartitionsDontHaveResources = false;
+          int resourceId = resources.iterator().next();
+          resourceIdToInstances.get(resourceId).addAll(instanceNames);
+          // Some instance names might not have a resource name, now assign the resource name to it
+          instanceNames.forEach(in -> instanceNameToOneResource.put(in, resourceId));
+          Set<String> s = resourceIdToInstances.get(resourceId);
+          ensureOrThrow(s.size() <= maxInstancesInOneResourceForFullAuto,
+              "Resource has more than " + maxInstancesInOneResourceForFullAuto + " hosts: " + s);
+        } else {
+          String errorMessage =
+              String.format("Partition %s belong to different hosts of different resources, hosts: %s, resources %s",
+                  ent.getKey(), instanceNames, resources);
+          ensureOrThrow(false, errorMessage);
+        }
+      }
+      // Run this logic for two rounds so that we can use the later partitions to exam prior partitions
```

Comment:
Me too :)

Reply:
here is the one example:
Assume we have this partition layout:
`{
    "partition1": {"host1", "host2", "host3" },
    "partition2": {"host2", "host3", "host4"}
}`

And we know that "host1" belong to resource1, but we don't know about host2, host3, and host4. Now when we iterate through this map, we encountered "partition2" before "partition1". Since host1 belong to resource1, and host2, host3 share partition1 with host1, so they belong to resource1 as well. So as host4. But when we first iterate through partition2, we don't know that. We have to iterate through partition1 in order to know that. 

That's why need two rounds of iteration. 

If the first round, we missed host2, host3, host4's resource, at the second round, we would know that they all belong to resource 1.